app/src/main/java/com/demoapp/database/DataRepository.kt

```diff
@@ -1,71 +1,73 @@
+import com.demoapp.adapterutils.RecyclerViewType
 import com.demoapp.model.CategoryEntity
 import com.demoapp.model.DataEntity
 import com.demoapp.model.DataResponse
 import com.demoapp.retrofit.RetrofitInterface
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.withContext
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.runBlocking
 import javax.inject.Inject
 
 class DataRepository @Inject constructor(
     private val dataDuo: DataDuo,
     private val mService: RetrofitInterface
 ) {
 
-    suspend fun getLocalData(): List<DataEntity> {
+    fun getLocalData(): Flow<List<RecyclerViewType>> = runBlocking {
+
         val categories = dataDuo.getAllCategories()
-        val dataList = dataDuo.getData()
+        val data = dataDuo.getData()
 
-        return categories.flatMap {
-            listOf(
-                DataEntity(
-                    name = "",
-                    description = "",
-                    price = 0.0,
-                    image_link = "",
-                    product_id = "",
-                    category_name = it.name,
-                    is_category = true
+        return@runBlocking categories.combine(data) { category, dataEntity ->
+            category.flatMap {
+                listOf(
+                    CategoryViewRM(it.name)
+                ).plus(
+                    dataEntity.filter { dataEntity ->
+                        dataEntity.category_name == it.name
+                    }.map { dataEntity ->
+                        DataViewRM(
+                            dataEntity.name, dataEntity.image_link
+                        )
+                    }
                 )
-            ).plus(dataList.filter { dataEntity ->
-                dataEntity.category_name == it.name
-            })
+            }
         }
     }
 
-    suspend fun getCategoryData() {
+    suspend fun getCategoryData(): DataResponse? {
         val response = mService.getCategoryWithData("89359ade-c88d-4048-815d-2e1e652728f7")
-
-        withContext(Dispatchers.Main) {
-            if (response.isSuccessful) {
-                response.body()?.let { insertCategoryData(it) }
-            }
+        if (response.isSuccessful) {
+            return response.body()
         }
+        return null
     }
 
-    private suspend fun insertCategoryData(response: DataResponse) {
+    suspend fun insertData(response: DataResponse) {
         if (!response.isNullOrEmpty()) {
-            dataDuo.removeCategoryData()
-            dataDuo.removeData()
 
-            response.forEach { category ->
-                dataDuo.insertCategories(CategoryEntity(name = category.category))
+            dataDuo.insertAllCategories(response.map {
+                CategoryEntity(name = it.category)
+            })
 
-                category.items.forEach {
-                    dataDuo.insertData(
+            response.map {
+                dataDuo.insertAllSubCategories(
```

Comment:
Why we need insert inside map ?

Reply:
Using map, app creates model to be inserted to local db from api response
Creating **CategoryEntity** from **category** field in response
Creating **DataEntity** from response with **categoryName** to manage reference between 2 tables

Reply:
Removed this part and managed through data source